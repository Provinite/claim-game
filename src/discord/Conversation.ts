import { APIMessageContentResolvable, Channel, Message } from "discord.js";
import { awaitNextMessage } from "./awaitNextMessage";

export class Conversation {
  constructor(private message: Message) {}
  say(msg: APIMessageContentResolvable): Promise<Message> {
    return this.message.channel.send(msg);
  }
  reply(msg: APIMessageContentResolvable): Promise<Message> {
    return this.message.reply(msg);
  }

  async listen(): Promise<Message> {
    return awaitNextMessage(
      this.message.client,
      (m) =>
        m.channel.id === this.message.channel.id &&
        m.author.id === this.message.author.id
    );
  }

  async askForChannel(prompt: APIMessageContentResolvable): Promise<Channel> {
    while (true) {
      await this.say(prompt);
      const response = await this.listen();
      const channelMentions = Array.from(response.mentions.channels);
      if (channelMentions.length) {
        return channelMentions[0][1];
      }
    }
  }
}
