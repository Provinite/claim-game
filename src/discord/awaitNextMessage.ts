import { Client, Message } from "discord.js";

export const awaitNextMessage = (
  client: Client,
  filter: (msg: Message) => boolean
) => {
  return new Promise<Message>((res, rej) => {
    const listener = (msg: Message) => {
      if (filter(msg)) {
        client.removeListener("message", listener);
        res(msg);
      }
    };
    client.on("message", listener);
  });
};
