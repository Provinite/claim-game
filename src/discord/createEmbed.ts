import { MessageEmbed } from "discord.js";

export function createEmbed() {
  return new MessageEmbed()
    .setColor("#DBA901")
    .setAuthor(
      "Prov",
      "https://avatars.githubusercontent.com/u/13734094?s=400&u=09327218ed4a22862c6d1ce9e68116f7a26d5ada&v=4",
      "https://github.com/provinite/claim-game"
    );
}
