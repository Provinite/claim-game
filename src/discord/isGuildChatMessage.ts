import { Guild, GuildMember, Message, TextChannel } from "discord.js";

export const isGuildChatMessage = (
  msg: Message
): msg is Message & {
  guild: Guild;
  member: GuildMember;
  channel: TextChannel;
} => {
  return Boolean(msg.member && msg.guild && msg.channel);
};
