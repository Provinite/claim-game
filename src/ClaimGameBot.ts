import { Client, Message } from "discord.js";
import { processBugCommand } from "./commands/BugCommand";
import { processClaimCommand } from "./commands/ClaimCommand";
import { processCreditsCommand } from "./commands/CreditsCommand";
import { processPotentialFulfillmentCommand } from "./commands/FulfillmentCommand";
import { processHelpCommand } from "./commands/HelpCommand";
import { processInitCommand } from "./commands/InitCommand";
import { processMyClaimsCommand } from "./commands/MyClaimsCommand";
import { processSettingsCommand } from "./commands/SettingsCommand";
import { isGuildChatMessage } from "./discord/isGuildChatMessage";
import { getGuildSettings } from "./guildSettings/GuildSettingsCache";
import { getAllGuildSettings } from "./guildSettings/GuildSettingsService";

const commands: Record<string, (msg: Message) => void> = {
  init: processInitCommand,
  claim: processClaimCommand,
  myclaims: processMyClaimsCommand,
  settings: processSettingsCommand,
  help: processHelpCommand,
  credits: processCreditsCommand,
  bug: processBugCommand,
};

export class ClaimGameBot {
  private commandPattern!: RegExp;
  constructor(private client: Client) {}

  async start() {
    const TOKEN = process.env.cc_claim_game_discord_token;
    // fetch all guild settings, priming the guild settings cache
    await getAllGuildSettings();

    this.client.on("ready", () => this.handleClientReady());
    this.client.on("message", (msg) => this.handleMessage(msg));

    await this.client.login(TOKEN);
    console.log("Logged in!");
  }
  stop() {
    const iv = (this.client as any)._intervals as Set<NodeJS.Timeout>;
    iv.forEach((t) => this.client.clearInterval(t));
    this.client.destroy();
  }
  async handleMessage(msg: Message) {
    if (msg.author.bot) {
      return;
    }

    if (!this.commandPattern) {
      return;
    }

    const hasCommand = this.commandPattern.test(msg.content);
    if (hasCommand || msg.channel.type === "dm") {
      let msgContent = msg.content;

      if (hasCommand) {
        msgContent = msgContent.substr(msg.content.indexOf(">") + 1);
      }
      msgContent = msgContent.trim();
      let firstWord = msgContent.split(/\b/)[0];
      if (firstWord) {
        firstWord = firstWord.toLowerCase();
      } else {
        return;
      }
      if (firstWord.length > 100) {
        return;
      }
      const cmd = commands[firstWord];
      if (cmd) {
        await cmd(msg);
      }
    } else if (isGuildChatMessage(msg)) {
      const fulfillmentChannelId = getGuildSettings(
        msg.guild.id
      )?.fulfillmentChannelId;
      if (msg.channel.id === fulfillmentChannelId) {
        await processPotentialFulfillmentCommand(msg);
      }
    }
  }

  async handleClientReady() {
    this.commandPattern = new RegExp(`^<@!?${this.client.user!.id}>`);
  }
}
