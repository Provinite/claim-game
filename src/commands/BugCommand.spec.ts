import { Client, SnowflakeUtil } from "discord.js";
import { MockedObjectDeep } from "ts-jest/dist/utils/testing";
import { ClaimGameBot } from "../ClaimGameBot";
import { createMockTextChannel } from "../test/discord-mocks/Channel.mock";
import { createMockClient } from "../test/discord-mocks/Client.mock";
import { createMockGuild } from "../test/discord-mocks/Guild.mock";
import { createMockCommandMessage } from "../test/discordHelpers";

describe("command:BugCommand", () => {
  let bot: ClaimGameBot;
  let mockClient: MockedObjectDeep<Client>;
  beforeEach(async () => {
    mockClient = createMockClient();
    mockClient.user = { id: SnowflakeUtil.generate() } as any;

    bot = new ClaimGameBot(mockClient);
    await bot.start();
    mockClient.emit("ready");
  });

  afterEach(async () => await bot.stop());

  it("replies with a link to github issues", async () => {
    if (!mockClient.user) throw new Error("Client needs to be defined");

    const message = createMockCommandMessage(
      "bug",
      mockClient,
      createMockTextChannel(createMockGuild(mockClient))
    );
    message.author = { bot: false } as any;
    jest.spyOn(message, "reply").mockResolvedValue(null as any);

    await bot.handleMessage(message);

    expect(message.reply.mock.calls[0][0]).toMatchSnapshot();
  });
});
