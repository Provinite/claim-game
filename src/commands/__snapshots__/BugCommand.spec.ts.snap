// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command:BugCommand replies with a link to github issues 1`] = `
Object {
  "author": Object {
    "icon_url": "https://avatars.githubusercontent.com/u/13734094?s=400&u=09327218ed4a22862c6d1ce9e68116f7a26d5ada&v=4",
    "name": "Prov",
    "url": "https://github.com/provinite/claim-game",
  },
  "color": 6185895,
  "description": "Please report all issues on the github issues page at https://www.github.com/provinite/claim-game/issues",
  "fields": Array [],
  "footer": Object {
    "icon_url": undefined,
    "text": "We appreciate all reports and suggestions!",
  },
  "image": null,
  "thumbnail": null,
  "timestamp": null,
  "title": "Bug Reports",
  "type": "rich",
  "url": null,
}
`;
